packages/material-ui-lab/src/TreeView/TreeView.js

```diff
@@ -394,164 +523,218 @@ const TreeView = React.forwardRef(function TreeView(props, ref) {
   /*
    * Mapping Helpers
    */
-
-  const addNodeToNodeMap = (id, childrenIds) => {
-    const currentMap = nodeMap.current[id];
-    nodeMap.current[id] = { ...currentMap, children: childrenIds, id };
-
-    childrenIds.forEach((childId) => {
-      const currentChildMap = nodeMap.current[childId];
-      nodeMap.current[childId] = { ...currentChildMap, parent: id, id: childId };
+  const registerNode = React.useCallback((node) => {
+    const { id, index, parentId, expandable } = node;
+    setNodeMap((oldMap) => {
+      oldMap[id] = { id, index, parentId, expandable };
+      return oldMap;
     });
-  };
-
-  const getNodesToRemove = React.useCallback((id) => {
-    const map = nodeMap.current[id];
-    const nodes = [];
-    if (map) {
-      nodes.push(id);
-      if (map.children) {
-        nodes.concat(map.children);
-        map.children.forEach((node) => {
-          nodes.concat(getNodesToRemove(node));
-        });
-      }
-    }
-    return nodes;
   }, []);
 
-  const cleanUpFirstCharMap = React.useCallback((nodes) => {
-    const newMap = { ...firstCharMap.current };
-    nodes.forEach((node) => {
-      if (newMap[node]) {
-        delete newMap[node];
-      }
-    });
-    firstCharMap.current = newMap;
-  }, []);
-
-  const removeNodeFromNodeMap = React.useCallback(
+  const unregisterNode = React.useCallback(
     (id) => {
-      const nodes = getNodesToRemove(id);
```

Comment:
So much cleaner. Each node only cares about itself now and we get rid of a lot of looping :)